creusot-contracts/src/std/ptr.rs

```diff
@@ -81,6 +84,75 @@ impl<T: ?Sized> PointerExt<T> for *mut T {
     }
 }
 
+pub trait SizedPointerExt<T> {
+    /// Logic version of `add` for pointers.
+    #[logic]
+    fn offset_logic(self, offset: Int) -> RawPtr<T>;
+
+    /// Restriction of `add` that requires evidence that the addition is safe.
+    /// See contracts in implementations below.
+    ///
+    /// From https://doc.rust-lang.org/std/primitive.pointer.html#method.add:
+    ///
+    /// > If any of the following conditions are violated, the result is Undefined Behavior:
+    /// > - The offset in bytes, `count * size_of::<T>()`, computed on mathematical
+    /// >   integers (without “wrapping around”), must fit in an `isize`.
+    /// > - If the computed offset is non-zero, then `self` must be derived from a
+    /// >   pointer to some allocated object, and the entire memory range between
+    /// >   `self` and the result must be in bounds of that allocated object.
+    /// >   In particular, this range must not “wrap around” the edge of the address space.
+    unsafe fn add_own(self, offset: usize, own_offset: Ghost<&PtrOwn<T>>) -> Self;
+}
+
+extern_spec! {
+    mod creusot_contracts {
+        mod std {
+            mod ptr {
+                trait SizedPointerExt<T> {
+                    #[requires(self.offset_logic(offset@) == own_offset.ptr())]
+                    unsafe fn add_own(self, offset: usize, own_offset: Ghost<&PtrOwn<T>>) -> Self;
+                }
+            }
+        }
+    }
+}
+
+impl<T> SizedPointerExt<T> for *const T {
+    #[trusted]
+    #[logic]
+    #[open(self)]
+    fn offset_logic(self, offset: Int) -> RawPtr<T> {
+        let _ = offset;
+        dead
+    }
+
+    // TODO: The offset in bytes, `count * size_of::<T>()`, must fit in an `isize`.
+    #[trusted]
+    #[requires(self.offset_logic(offset@) == own_offset.ptr())]
+    #[ensures(self.offset_logic(offset@) == result)]
+    unsafe fn add_own(self, offset: usize, own_offset: Ghost<&PtrOwn<T>>) -> Self {
+        self.add(offset)
+    }
+}
+
+impl<T> SizedPointerExt<T> for *mut T {
+    #[trusted]
+    #[logic]
+    #[open(self)]
+    fn offset_logic(self, offset: Int) -> RawPtr<T> {
+        let _ = offset;
+        dead
+    }
+
+    // TODO: The offset in bytes, `count * size_of::<T>()`, must fit in an `isize`.
+    #[trusted]
+    #[requires(self.offset_logic(offset@) == own_offset.ptr())]
+    // #[ensures(result as RawPtr<T> == self.offset_logic(offset))] // TODO: cast *mut to RawPtr ?
```

Comment:
Is this forbidden in Pearlite? We should allow these casts, and translate them by identity (using the Coerce construct).